telemetry/convert.py

```diff
@@ -146,10 +149,127 @@ def convert_json(self, jsonstr, date, ip=None):
+                info["isGPU2Active"] = bool(adapters[1].get("GPUActive"))
+
+        # Fix up clientID
+        json_dict["clientID"] = json_dict.get("clientId")
+
+        # Step 3:
+        #   Add a "ver" field to make it look like a classic ping, and
+        #   delete the "version" field to make the conversion idempotent.
+        #   This ping will then be passed through convert_obj again to
+        #   fix up the histograms.
+        json_dict["original.version"] = json_dict["version"]
+        del json_dict["version"]
+        json_dict["ver"] = Converter.VERSION_UNCONVERTED
+
     def convert_obj(self, json_dict, date, ip=None):
-        info = json_dict.get("info", None)
-        # Check if the payload is already converted:
-        if "ver" in json_dict:
+        if "version" in json_dict:
+            # This is a new unified (FHR+Telemetry) ping
+            if json_dict["version"] != Converter.VERSION_UNIFIED:
+                raise ValueError("Unknown unified ping version: " + str(json_dict["version"]))
+            elif "type" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'type' field")
+            elif "application" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'application' field")
+            elif "payload" not in json_dict:
+                raise ValueError("Unified ping has no top-level 'payload' field")
+
+            # Create a ping.info field and copy select ping.application fields.
+            # This is the minimum required for compatibility with the calling
+            # process_incoming* scripts
+            info = json_dict["info"] = {}
+            self.add_info_fields(info, json_dict["application"], appFieldMap)
+            if json_dict["payload"].get("info"):
+                info["reason"] = json_dict["payload"]["info"].get("reason")
+
+            if json_dict["type"] == "main":
+                # Handle "Main" ping type
+                if "environment" not in json_dict:
+                    raise ValueError("Unified ping with type=main has no 'environment' field")
+
+                if info.get("reason") == "saved-session":
+                    # This is a saved-session ping in the new unified format,
+                    # make it look like a classic saved-session ping
+                    self.convert_saved_session(json_dict)
+                    # call convert again to convert the histograms
+                    return self.convert_obj(json_dict, date, ip)
+                else:
+                    # Most likely a sub-session daily or environment-changed ping, so don't alter it further,
+                    # just let it get written out
+                    pass
+            else:
+                # Some new ping type: activation/upgrade/deletion, or future types.
+                # Don't alter it further
+                pass
+
+        elif "ver" in json_dict:
```

Comment:
Move this up to be the first check - if a ping has a top-level `ver` field, we should handle it the old way. Then if any existing/old ping also has a `version` field (I don't think any do, but just in case), we will still process it the same way.